crates/matrix-sdk-base/src/read_receipts.rs

```diff
@@ -156,6 +281,127 @@ impl PreviousEventsProvider for () {
+            // Note: by using a lax comparison here, we properly handle the case where we
+            // received events that we have already seen with a persisted read
+            // receipt.
+            if event_pos >= *best_pos {
+                *best_pos = event_pos;
+                self.best_receipt = Some(event_id.to_owned());
+                trace!("saving better");
+            }
+        } else {
+            // We didn't have a previous receipt, this is the first one we
+            // store: remember it.
+            self.best_pos = Some(event_pos);
+            self.best_receipt = Some(event_id.to_owned());
+            trace!("saving for the first time");
+        }
+    }
+
+    /// Try to match pending receipts against new events.
+    fn handle_pending_receipts(&mut self, pending: &mut RingBuffer<OwnedEventId>) {
+        // Try to match stashed receipts against the new events.
+        trace!("handle_pending_receipts");
+        pending.retain(|event_id| {
+            if let Some(event_pos) = self.event_id_to_pos.get(event_id) {
+                // Maybe select this read receipt as it might be better than the ones we had.
+                self.try_select_better(event_id, *event_pos);
+
+                // Remove this stashed read receipt from the pending list, as it's been
+                // reconciled with its event.
+                false
+            } else {
+                // Keep it for further iterations.
+                true
+            }
+        });
+    }
+
+    /// Try to match new receipts against all (new and old) events.
+    ///
+    /// Returns all the new pending receipts (those for which we didn't have a
+    /// known matching event).
+    fn handle_new_receipt(
+        &mut self,
+        user_id: &UserId,
+        receipt_event: &ReceiptEventContent,
+    ) -> Vec<OwnedEventId> {
+        trace!("handle_new_receipt");
+        let mut pending = Vec::new();
+        // Now consider new receipts.
+        for (event_id, receipts) in &receipt_event.0 {
+            for ty in [ReceiptType::Read, ReceiptType::ReadPrivate] {
+                if let Some(receipt) = receipts.get(&ty).and_then(|receipts| receipts.get(user_id))
+                {
+                    if matches!(receipt.thread, ReceiptThread::Main | ReceiptThread::Unthreaded) {
+                        if let Some(event_pos) = self.event_id_to_pos.get(event_id) {
+                            self.try_select_better(event_id, *event_pos);
+                        } else {
+                            // It's a new pending receipt.
+                            pending.push(event_id.clone());
+                        }
+                    }
+                }
+            }
+        }
+        pending
+    }
+
+    /// Returns a *new* better read receipt. If it's not set, we can consider
+    /// that each new event is *after* the previous read receipt.
+    fn finish(self) -> Option<LatestReadReceipt> {
```

Comment:
We could name this `select` - it might make its usage in `compute_notifications` easier to understand.